### DISCUSSION TITLE
Old-school CGI guestbook in Joker

### DISCUSSION THREAD

**Discussion by dbohdan** [candid82/joker#503] [Show and tell]
I have made a thing using Joker: [joker-cgi-guestbook](https://github.com/dbohdan/joker-cgi-guestbook). It is a toy old-school [CGI](https://en.wikipedia.org/wiki/Common_Gateway_Interface) guestbook that stores the guestbook entries in Bolt and uses `joker.crypto/hmac` for a stateless captcha (meaning that individual captchas aren't tracked on the server).

You can try it live at [dbohdan.sdf.org/guestbook/](http://dbohdan.sdf.org/guestbook/).

<img alt="A screenshot of the guestbook." src="https://github.com/candid82/joker/assets/3179832/de65cafc-9c24-4d7b-8d2d-039b123c7935" width="320">

**Comment by candid82** [comment]
@dbohdan This is pretty cool! The implementation is truly in the spirit of Joker: simple and concise.

**Reply by dbohdan** [reply]
Thanks! I wanted the guestbook to be self-contained like Joker. At first I thought I'd have to rely on an external signing tool like [minisign](https://github.com/jedisct1/minisign) but then found HMAC in the standard library.

**Comment by candid82** [comment]
@dbohdan I noticed you had to implement [parse-query](https://github.com/dbohdan/joker-cgi-guestbook/blob/master/index.joke#L192) function. I just added joker.url/parse-query to the standard library to make things like this easier: https://github.com/candid82/joker/commit/4f0c27f4998aa2bc921d2a80c41c52b1b094df25

**Reply by dbohdan** [reply]
This helps, thanks! I have refactored the code to use it: https://github.com/dbohdan/joker-cgi-guestbook/commit/912fd897b92bf4d97a4cb621012fcb45fc7493fc. It seems to work fine: http://dbohdan.sdf.org/guestbook/#entry-33d9cf.

(Check out the new [entry ids](https://github.com/dbohdan/joker-cgi-guestbook/commit/8503f10e287f2da33833492d5c3e702478966811), like the one I have just linked to. They also use HMAC so an attacker won't know how many entries you have deleted.)
